### PR TITLE
Relay action and snmp configuration support for Coral PDU

### PR DESCRIPTION
**Pull request by cxhong**
Add supports for coral pdu
https://github.com/xcat2/xcat-core/wiki/Coral-PDU-supports

rpower command:
````
rpower coralpdu stat
rpower coralpdu off
rpower coralpdu on
rpower coralpdu relay=2 off
rpower coralpdu relay=3 on
````

rspconfig command:
````
rspconfig coralpdu snmpcfg
````
snmp configuration needs to predefine in the pdu table:
````
# tabdump pdu
#node,nodetype,pdutype,outlet,username,password,snmpversion,community,snmpuser,authtype,authkey,privtype,privkey,seclevel,comments,disable
"coral-pdu","pdu","crpdu",,,,"3","xcatpass","xcatadm",,"xcatpassword",,,"authPriv",,
````

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/pdu.pm [hunk 1]
```diff
@@ -382,13 +441,60 @@ sub powerstat {
...
+
+                    for (my $relay = 1; $relay <= 3; $relay++) {
+                        relaystat($pdu, $relay, $snmpcmd, $callback);
+                    }
+                    next;
+                }
+           } 
+            xCAT::SvrUtils::sendmsg("please config snmpv3 to be able to query pdu relay status", $callback,$pdu);
+            xCAT::SvrUtils::sendmsg("    use chdef command to add pdu snmpv3 attributes to pdu table", $callback);
+            xCAT::SvrUtils::sendmsg("    then run 'rspconfig $pdu snmpcfg' command ", $callback);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/pdu.pm]: No $pdu specified for the next 2 messages, will it be a problem?

File: xCAT-server/lib/xcat/plugins/pdu.pm [hunk 2]
```diff
@@ -844,5 +1000,264 @@ sub showMonitorData {
...
+            relay_action($session, $pdu, $relay, "ON", $callback);
+        } elsif ( $subcmd eq "reset") {
+            relay_action($session, $pdu, $relay, "OFF", $callback);
+            relay_action($session, $pdu, $relay, "ON", $callback);
+        } else {
+            xCAT::SvrUtils::sendmsg(" subcmd $subcmd is not support", $callback,$pdu);
+        }
+        $relay_num++;
+    }
+    #$session->hard_close();
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/pdu.pm]: Why not close the ssh session?

File: perl-xCAT/xCAT/Schema.pm [hunk 3]
```diff
@@ -694,17 +694,25 @@ passed as argument rather than by table value',
...
+            username => 'The remote login user name',
+            password => 'The remote login password',
+            snmpversion => 'The version to use to communicate with switch.  SNMPv1 is assumed by default.',
+            community => 'The community string to use for SNMPv1/v2',
+            snmpuser => 'The username to use for SNMPv3 communication, ignored for SNMPv1',
+            authtype => 'The authentication protocol(MD5|SHA) to use for SNMPv3.',
+            authkey => 'The authentication passphrase for SNMPv3 ',
+            privtype => 'The privacy protocol(AES|DES) to use for SNMPv3.',
+            privkey => 'The privacy passphrase to use for SNMPv3.',
+            seclevel => 'The Security Level(noAuthNoPriv|authNoPriv|authPriv) to use for SNMPv3.',
```
[the remark this answers is not in the document]
Comment (cxhong) [reply perl-xCAT/xCAT/Schema.pm]: community string is only for snmpv1/2,  it apply for both crpdu and other type of PDU.
crpdu only supports snmpv3 right now.  I don't think other type of PDU needs snmpv3 configuration.
we can use rinv to query machinetype, modlenum and serialnum.  don't think it needs in the PDU table.
The PDU table is setup up as switch table so far.

File: xCAT-server/lib/xcat/plugins/pdu.pm [hunk 4]
```diff
@@ -71,6 +71,41 @@ sub handled_commands
...
 
+sub pdu_usage
+{
+    my ($callback, $command) = @_;
+    my $usagemsg = 
+    "Usage: 
+        pdudiscover [<noderange>|--range ipranges] [-r|-x|-z] [-w] [-V|--verbose] [--setup]
+        rpower pdunodes [off|on|stat|reset]
+
+     The following command supports IR PDU with pdutype=irpdu :
```
[the remark this answers is not in the document]
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: The IR PDU (the one on our test system) has machinetype `1u`.
crpdu (coral pdu) and irpdu (other than coral pdu) are two types of PDU we are supported.  we verified the command for crpdu.  there are different kinds of irpdu too,  I only verified the one on our test system.  they are no available irpdu in the lab for us to do the test.

File: xCAT-server/lib/xcat/plugins/pdu.pm (hunk 1 above)
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: looks fine.  there are space on the front for next two lines.

File: xCAT-server/lib/xcat/plugins/pdu.pm [hunk 5]
```diff
@@ -844,5 +1000,264 @@ sub showMonitorData {
...
+        }
+    } else {
+        xCAT::SvrUtils::sendmsg(" This command is not support, please define relay number and action", $callback);
+    }
+
+}
+
+#-------------------------------------------------------
+
+=head3  process_powerrelay
```
[the remark this answers is not in the document]
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: oh, right

File: xCAT-server/lib/xcat/plugins/pdu.pm [hunk 6]
```diff
@@ -844,5 +1000,264 @@ sub showMonitorData {
...
+    foreach my $cmd (@exargs) {
+        if ($cmd =~ /=/ ) {
+            my ($key, $value) = split(/=/, $cmd);
+            $relay = $value;
+        } else {
+            $action = $cmd;
+        }
+    }
+    if ( (defined $relay) && (defined $action) ) {
+        my $relay_count = 1;
```
[the remark this answers is not in the document]
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: changed

File: xCAT-server/lib/xcat/plugins/pdu.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: modified

File: xCAT-server/lib/xcat/plugins/pdu.pm (hunk 2 above)
Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/pdu.pm]: yeh, should close the session.  modified

File: perl-xCAT/xCAT/data/switchinfo.pm [hunk 7]
```diff
@@ -41,7 +41,8 @@ our %global_switch_type = (
     MELLAN => "Mellanox",
     Cumulus => "onie",
     cumulus => "onie",
-    Edgecore => "onie"
+    Edgecore => "onie",
+    coral => "crpdu"
```
Comment (whowutwut) [inline perl-xCAT/xCAT/data/switchinfo.pm]: Should we standardize the upper lower case?  

Why two `cumulus` and `Cumulus` ? 

Should `coral` -> `Coral` instead ... or `Constellation`  :-)

Comment (cxhong) [reply perl-xCAT/xCAT/data/switchinfo.pm]: `coral => crpdu`,   
`crpdu` is the PDU type for Constellation PDU,  
`irpdu` is the PDU type for Intelligent PDU.
 
`coral` is used for match string from `snmpwalk -v1 -c public $pduname system` command

**Comment by zet809** [comment]
Looks good to me, agree to merge.

**Pull request merged**